File: app/src/main/java/com/example/uscatterbrain/network/LuidPacket.java
package com.example.uscatterbrain.network;

import com.example.uscatterbrain.ScatterProto;
import com.github.davidmoten.rx2.Bytes;
import com.google.protobuf.ByteString;
import com.google.protobuf.GeneratedMessageLite;
import com.goterl.lazycode.lazysodium.interfaces.GenericHash;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.UUID;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;

public class LuidPacket implements ScatterSerializable {
    private final ScatterProto.Luid mLuid;
    private final boolean isHashed;
    private UUID luidtag;

    private LuidPacket(Builder builder) {
        this.isHashed = builder.enablehash;
        if (builder.enablehash) {
            mLuid = ScatterProto.Luid.newBuilder()
                    .setValHash(ByteString.copyFrom(calculateHashFromUUID(builder.uuid)))
                    .build();
        } else {
            ScatterProto.UUID u = protoUUIDfromUUID(builder.uuid);
            mLuid = ScatterProto.Luid.newBuilder()
                    .setValUuid(u)
                    .build();
        }
    }

    private LuidPacket(InputStream is) throws IOException {
        mLuid = CRCProtobuf.parseFromCRC(ScatterProto.Luid.parser(), is);
        isHashed = mLuid.getValCase() == ScatterProto.Luid.ValCase.VAL_HASH;
    }

    private static ScatterProto.UUID protoUUIDfromUUID(UUID uuid) {
        return ScatterProto.UUID.newBuilder()
                .setLower(uuid.getLeastSignificantBits())
                .setUpper(uuid.getMostSignificantBits())
                .build();
    }

    private static UUID protoUUIDtoUUID(ScatterProto.UUID uuid) {
        return new UUID(uuid.getUpper(), uuid.getLower());
    }

    private static byte[] calculateHashFromUUID(UUID uuid) {
        byte[] hashbytes = new byte[GenericHash.BYTES];
        ByteBuffer uuidBuffer = ByteBuffer.allocate(16);
        uuidBuffer.putLong(uuid.getMostSignificantBits());
        uuidBuffer.putLong(uuid.getLeastSignificantBits());
        byte[] uuidbytes = uuidBuffer.array();
        LibsodiumInterface.getSodium().crypto_generichash(
                hashbytes,
                hashbytes.length,
                uuidbytes,
                uuidbytes.length,
                null,
                0
        );

        return hashbytes;
    }

    public byte[] getHash() {
        if (isHashed) {
            return mLuid.getValHash().toByteArray();
        } else {
            return new byte[0];
        }
    }

    public UUID getHashAsUUID() {
        byte[] h = mLuid.getValHash().toByteArray();
        if (h.length != GenericHash.BYTES) {
            return null;
        } else if (isHashed) {
            ByteBuffer buf = ByteBuffer.wrap(h);
            //note: this only is safe because crypto_generichash_BYTES_MIN is 16
            return new UUID(buf.getLong(), buf.getLong());
        } else {
            return null;
        }
    }

    public boolean verifyHash(LuidPacket packet) {
        if (packet.isHashed == this.isHashed) {
            return false;
        } else if (this.isHashed) {
            byte[] hash = calculateHashFromUUID(packet.getLuid());
            return Arrays.equals(hash, getHash());
        } else {
            byte[] hash = calculateHashFromUUID(this.getLuid());
            return Arrays.equals(hash, packet.getHash());
        }
    }

    public static Single<LuidPacket> parseFrom(InputStream inputStream) {
        return Single.fromCallable(() -> new LuidPacket(inputStream));
    }

    public static Single<LuidPacket> parseFrom(Observable<byte[]> flowable) {
        InputStreamObserver observer = new InputStreamObserver();
        flowable.subscribe(observer);
        return LuidPacket.parseFrom(observer).doFinally(observer::close);
    }

    public static Single<LuidPacket> parseFrom(Flowable<byte[]> flowable) {
        InputStreamFlowableSubscriber observer = new InputStreamFlowableSubscriber();
        flowable.subscribe(observer);
        return LuidPacket.parseFrom(observer).doFinally(observer::close);
    }

    public ScatterProto.Luid.ValCase getValCase() {
        return mLuid.getValCase();
    }

    @Override
    public byte[] getBytes() {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        try {
            CRCProtobuf.writeToCRC(mLuid, os);
            return os.toByteArray();
        } catch (IOException ignored) {
            return null;
        }
    }

    @Override
    public ByteString getByteString() {
        return ByteString.copyFrom(getBytes());
    }

    @Override
    public Completable writeToStream(OutputStream os) {
        return Completable.fromAction(() -> CRCProtobuf.writeToCRC(mLuid, os));
    }

    @Override
    public GeneratedMessageLite getMessage() {
        return mLuid;
    }

    @Override
    public Flowable<byte[]> writeToStream(int fragsize) {
        return Bytes.from(new ByteArrayInputStream(getBytes()), fragsize);
    }

    @Override
    public PacketType getType() {
        return PacketType.TYPE_LUID;
    }

    @Override
    public void tagLuid(UUID luid) {
        luidtag = luid;
    }

    public boolean isHashed() {
        return isHashed;
    }

    public UUID getLuid() {
        return protoUUIDtoUUID(mLuid.getValUuid());
    }

    public static Builder newBuilder() {
        return new Builder();
    }

    public static class Builder {
        private UUID uuid;
        private boolean enablehash;

        private Builder() {
            enablehash = false;
        }

        public Builder enableHashing() {
            this.enablehash = true;
            return this;
        }

        public Builder setLuid(UUID uuid) {
            this.uuid = uuid;
            return this;
        }

        public LuidPacket build() {
            if (uuid == null) {
                throw new IllegalArgumentException("uuid required");
            }

            return new LuidPacket(this);
        }
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/LuidPacket.kt
package net.ballmerlabs.uscatterbrain.network

import com.google.protobuf.ByteString
import com.goterl.lazysodium.interfaces.GenericHash
import net.ballmerlabs.uscatterbrain.ScatterProto.Luid
import net.ballmerlabs.uscatterbrain.ScatterProto.Luid.hashed
import net.ballmerlabs.uscatterbrain.db.hashAsUUID
import net.ballmerlabs.uscatterbrain.network.bluetoothLE.BluetoothLERadioModuleImpl
import java.nio.ByteBuffer
import java.util.UUID

fun getHashUuid(uuid: UUID?): UUID? {
    return if(uuid == null)
        null
    else
        BluetoothLERadioModuleImpl.bytes2uuid(LuidPacket.calculateHashFromUUID(uuid))
}

/**
 * Wrapper class for Luid protobuf message
 * @property isHashed true if this packet has only a hash value
 * @property luidVal identifier of remote peer
 * @property hash hash of this packet
 * @property protoVersion Scatterbrain protocol version (used for version checks)
 * @property hashAsUUID hash of this packet represented as UUID
 */
class LuidPacket (packet: Luid) : ScatterSerializable<Luid>(packet) {
    val isHashed: Boolean
    get() = packet.valCase == Luid.ValCase.VAL_HASH

    val luidVal: UUID = protoUUIDtoUUID(packet.valUuid)
    
    val hash: ByteArray
        get() = if (isHashed) {
            packet.valHash.toByteArray()
        } else {
            calculateHashFromUUID(luidVal)
        }

    val protoVersion: Int
        get() = if (!isHashed) {
            -1
        } else {
            packet.valHash.protoversion
        }

    //note: this only is safe because crypto_generichash_BYTES_MIN is 16
    private val hashAsUUID: UUID
        get() {
            val h = packet.valHash.hash.toByteArray()
            return if (!isHashed)
                getHashUuid(luidVal)!!
            else
                hashAsUUID(h)
        }

    fun verifyHash(p: LuidPacket): Boolean {
        return when {
            p.isHashed == isHashed -> {
                false
            }
            isHashed -> {
                val hash = calculateHashFromUUID(p.luidVal)
                hash.contentEquals(hash)
            }
            else -> {
                val hash = calculateHashFromUUID(luidVal)
                hash.contentEquals(p.hash)
            }
        }
    }

    override val type: PacketType
        get() = PacketType.TYPE_LUID

    data class Builder(
            var uuid: UUID? = null,
            var enablehash: Boolean = false,
            var version: Int = -1
    ) {
        fun enableHashing(protoversion: Int) = apply {
            enablehash = true
            version = protoversion
        }

        fun setLuid(uuid: UUID?) = apply {
            this.uuid = uuid
        }

        fun build(): LuidPacket {
            requireNotNull(uuid) { "uuid required" }
            val p = if (enablehash) {
                val h = hashed.newBuilder()
                        .setHash(ByteString.copyFrom(calculateHashFromUUID(uuid!!)))
                        .setProtoversion(version)
                        .build()
                Luid.newBuilder()
                        .setValHash(h)
                        .build()
            } else {
                val u = protoUUIDfromUUID(uuid!!)
                Luid.newBuilder()
                        .setValUuid(u)
                        .build()
            }
            return LuidPacket(p)
        }

    }

    companion object {
        fun calculateHashFromUUID(uuid: UUID): ByteArray {
            val hashbytes = ByteArray(GenericHash.BYTES)
            val uuidBuffer = ByteBuffer.allocate(16)
            uuidBuffer.putLong(uuid.mostSignificantBits)
            uuidBuffer.putLong(uuid.leastSignificantBits)
            val uuidbytes = uuidBuffer.array()
            LibsodiumInterface.sodium.crypto_generichash(
                    hashbytes,
                    hashbytes.size,
                    uuidbytes,
                    uuidbytes.size.toLong(),
                    null,
                    0
            )
            return hashbytes
        }

        fun newBuilder(): Builder {
            return Builder()
        }

        class Parser : ScatterSerializable.Companion.Parser<Luid, LuidPacket>(Luid.parser())
        fun parser(): Parser {
            return Parser()
        }
    }
}